test/testexprengine.cpp

```diff
@@ -448,6 +454,73 @@ class TestExprEngine : public TestFixture {
                       expr("void foo(const int *x) { if (f1() && *x > 12) dostuff(*x == 5); }", "=="));
     }
 
+    void ifAlwaysTrue1() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (1) {\n"
+                            "    a = 0;\n"
+                            "  }\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(distinct 1 0)\n"
+                                "(= 0 0)\n"
+                                "z3::sat\n";
+        ASSERT_EQUALS(expected, expr(code, "=="));
+    }
+
+    void ifAlwaysTrue2() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (12.3) {\n"
+                            "    a = 0;\n"
+                            "  }\n"
+                            "  return a == 0;\n"
+                            "}";
+        const char expected[] = "(distinct 12.3 0.0)\n"
+                                "(= 0 0)\n"
+                                "z3::sat\n";
+        ASSERT_EQUALS(expected, expr(code, "=="));
+    }
+
+    void ifAlwaysTrue3() {
+        const char code[] = "int foo() {\n"
+                            "  int a = 42;\n"
+                            "  if (\"test\") {\n"
+                            "    a = 0;\n"
+                            "  }\n"
+                            "  return a == 0;\n"
+                            "}";
+        // String literals are always true. z3 will not be involved.
+        ASSERT_EQUALS("", expr(code, "=="));
```

Comment:
hmm.. I am surprised that the output from expr is empty. but we can leave it like this for now.

Reply:
This is because I don't call z3 when I see a string literal: https://github.com/danmar/cppcheck/pull/2969/files#diff-798cacc2b750d79fa5159a692333c8e770ed78ce4eb66b9efde24d799173ce9eR2541

But it interesting, why `return a == 0` did not generate an equation.

Reply:
Yes the expr should output something for `a == 0`